dotnet/sdk-phase-1.md

```diff
@@ -0,0 +1,1052 @@
+
+    public void AddActivities(params object[] activities) { /*...*/ }
+
+    public void AddActivityDefinition(ActivityDefinition definition) { /*...*/ }
+
+    // Many more options not listed
+}
+
+public class TemporalWorker
+{
+    public TemporalWorker(IWorkerClient client, TemporalWorkerOptions options) { /*...*/ }
+
+    public Task ExecuteAsync(CancellationToken stoppingToken) { /*...*/ }
+}
+```
+
+Notes:
+
+* Like other options, technically all worker options are optional at compile time but we fail at runtime if task queue
+  is not set
+* Activity definitions are the underlying type of the activity collection (they have just a name and method to call) but
+  most users will call `AddActivities`
+* `AddActivities` intentionally take types or instances
+  * For types, all `[Activity`s must be on static methods, no instance methods allowed
+    * It was decided to do `AddActivities(typeof(MyStaticClass))` for all static method things than to have
+    `AddActivities<MyStaticClass>()` overload. All-static activities may not be that common and adding overloads is
+    unnecessary.
+  * For instances, `[Activity`s can be on static or instance methods
+  * Inheritance rules apply from the earlier activity definition section
+  * Users can use the lower-level `AddActivityDefinition` but we left this name unwieldy on purpose, we prefer
+    `AddActivities`
+* Since workflows are not currently implemented, at least one activity is required
+* Activities can be sync or async
+  * For future `Workflow.ExecuteActivity` uses, we have confirmed .NET can disambiguate in an overload between `Task`
+    return type and generic return type with no constraints
+  * Async activities are executed using the normal scheduler
+    * No need yet for a custom scheduler
+  * Sync activities are executed with `ThreadPool`
+    * No need yet for a custom executor
+* As much validation and non-async initialization is done in the constructor as possible
+* While it may be tempting to make `Worker` implement `IAsyncDisposable`, we want users to explicitly handle fatal
+  errors out of `ExecuteAsync`
+  * We can't throw out of `await using` blocks
+* Notice a lack of `StartAsync`, we want users to handle fatal errors from `ExecuteAsync`
+* Notice a lack of `ShutdownAsync`, we want users to use the cancellation token
+* `ExecuteAsync(CancellationToken)` was specifically done this way to match
+  `Microsoft.Extensions.Hosting.BackgroundService.ExecuteAsync`
+
+## Replayer
+
+Looks like:
+
+```csharp
+namespace Temporal.Worker;
+
+class WorkflowReplayer
+{
+    public WorkflowReplayer(TemporalWorkerOptions options) { /*...*/ }
+
+    public async Task<WorkflowReplayResult> ReplayWorkflow(
+        WorkflowHistory history, bool throwOnReplayFailure = true) { /*...*/ }
+
+#if NETCOREAPP3_0_OR_GREATER
+
+    public IAsyncEnumerator<WorkflowReplayResult> ReplayWorkflows(
+      IAsyncEnumerator<WorkflowHistory> histories, bool throwOnReplayFailure = true) { /*...*/ }
```

Comment:
Yes, technically the caller could write this whole method. It was decided on Python and others to provide an option to eagerly throw on first failure instead of return. No reason for .NET to deviate.